### PR TITLE
impl CURLOPT_CURLU

### PR DESCRIPTION
**Pull request by xquery**
Implement CURLOPT_CURLU (as per https://github.com/curl/curl/wiki/URL-API) so one may pass via curl_easy_setopt aka:

```
  CURL *handle;
  CURLU *urlp;
  CURLUcode uc;
  
  urlp = curl_url();
  uc = curl_url_set(urlp, CURLUPART_URL, "http://example.com/path/index.html", 0);

  curl_easy_setopt(handle, CURLOPT_CURLU, urlp);
  ...
```

this approach expands on the prior attempts ( #3128, #3160) ... I have opted to be simple (possibly stupid) vs clever - still unsure if further refactoring needed (eg. transfer.c:Curl_pretransfer).
 
I have attempted to document CURLOPT_CURLU and added docs/examples/urlapi.c to illustrate usage.

Comments, advice, rotten vegetables all gladly accepted.

### CODE REVIEW

File: docs/libcurl/opts/CURLOPT_CURLU.3 [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+.TH CURLOPT_CURLU 3 "28 Oct 2018" "libcurl 7.62.0" "curl_easy_setopt options"
+.SH NAME
+CURLOPT_CURLU \- set URL with CURLU *
+.SH SYNOPSIS
+#include <curl/curl.h>
+
+CURLcode curl_easy_setopt(CURL *handle, CURLOPT_CURLU, void *pointer);
+.SH DESCRIPTION
+Pass in a pointer to the \fIURL\fP to work with. The parameter should be a
+CURLU *.
```
Comment (bagder) [inline docs/libcurl/opts/CURLOPT_CURLU.3]: I think also emphasize that this option is then used *instead* of `CURLOPT_URL`. Also mention what happens if both options are set.

File: docs/libcurl/opts/CURLOPT_CURLU.3 [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  curl_easy_setopt(handle, CURLOPT_CURLU, urlp);
+
+  ret = curl_easy_perform(handle);
+
+  curl_url_cleanup(urlp);
+  curl_easy_cleanup(handle);
+}
+.fi
+.SH AVAILABILITY
+Added in 7.62.0.
```
Comment (bagder) [inline docs/libcurl/opts/CURLOPT_CURLU.3]: Added in 7.63.0 - the next version number allowing changes!

File: docs/libcurl/opts/CURLOPT_CURLU.3 [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+.SH RETURN VALUE
+Returns CURLE_OK if the option is supported, and CURLE_UNKNOWN_OPTION if not.
+.SH "SEE ALSO"
+.BR CURLOPT_URLBUFFERSIZE "(3),
```
Comment (bagder) [inline docs/libcurl/opts/CURLOPT_CURLU.3]: You meant `CURLOPT_URL` here, right? =)

File: docs/libcurl/symbols-in-versions [hunk 4]
```diff
@@ -372,6 +372,7 @@ CURLOPT_COOKIESESSION           7.9.7
 CURLOPT_COPYPOSTFIELDS          7.17.1
 CURLOPT_CRLF                    7.1
 CURLOPT_CRLFILE                 7.19.0
+CURLOPT_CURLU                   7.62.0
```
Comment (bagder) [inline docs/libcurl/symbols-in-versions]: Also likely to be 7.63.0

File: lib/transfer.c [hunk 5]
```diff
@@ -1348,17 +1348,24 @@ void Curl_init_CONNECT(struct Curl_easy *data)
...
   /* since the URL may have been redirected in a previous use of this handle */
   if(data->change.url_alloc) {
     /* the already set URL is allocated, free it first! */
     Curl_safefree(data->change.url);
     data->change.url_alloc = FALSE;
   }
+
+  if(!data->change.url && data->set.uh)
+    result = curl_url_get(data->set.uh,
+                        CURLUPART_URL, &data->set.str[STRING_SET_URL], 0);
```
Comment (bagder) [inline lib/transfer.c]: This should return if an error is returned from curl_url_get(), right?

Comment (xquery) [reply lib/transfer.c]: and should be CURLUcode uc instead of result ;)

File: docs/libcurl/symbols-in-versions (hunk 4 above)
Comment (xquery) [reply docs/libcurl/symbols-in-versions]: ack

File: docs/libcurl/opts/CURLOPT_CURLU.3 (hunk 3 above)
Comment (xquery) [reply docs/libcurl/opts/CURLOPT_CURLU.3]: trying to win the award for most number of errors in a commit

File: docs/libcurl/opts/CURLOPT_CURLU.3 (hunk 2 above)
Comment (xquery) [reply docs/libcurl/opts/CURLOPT_CURLU.3]: ack

**Comment by xquery** [comment]
now that we have passing tests, need to do another pass on the docs and example

**Comment by xquery** [comment]
frak,  in the future I will do git push --force but today I muddied up my branch .... closing this PR and reraising

**Pull request closed without merging**